### ISSUE TITLE
Class attrs added by metaclass __prepare__ incorrectly reports "not defined"

### ISSUE THREAD

**Issue by alkasm** [microsoft/pyright#1250]
**Describe the bug**
Pyright incorrectly reports that attributes added to the namespace in `__prepare__` in a metaclass are not defined inside a class definition which has that metaclass.

**To Reproduce**
See the code below. In VSCode with Pylance squiggles appear under `prepared_attr` and `decorator` inside the `Test` class definition.

**Expected behavior**
Pyright should not report errors here, since the variable is defined at this point.

**Screenshots or Code**
```python
class Meta(type):
    @classmethod
    def __prepare__(cls, *args):
        return {"decorator": lambda f: f, "prepared_attr": "prepared_attr_value"}


class Test(metaclass=Meta):
    prep = prepared_attr

    @decorator
    def f(self):
        return self.__class__.prepared_attr


t = Test()
print(Test.prep)
print(t.f())
```

**VS Code extension or command-line**
Pylance version 2020.12.0

**Additional context**
From the [Python metaclass docs](https://docs.python.org/3/reference/datamodel.html#metaclasses):

> When a class definition is executed, the following steps occur:
> 
> - MRO entries are resolved;
> - the appropriate metaclass is determined;
> - the class namespace is prepared;
> - the class body is executed;
> - the class object is created.

The namespace is constructed with `__prepare__` on the metaclass, then the class body is executed, and then the class itself is created. That means that when the class body is executed, anything added to the namespace inside `__prepare__` is scoped and available to use inside the class definition. However, pyright reports that variables added to the namespace in `__prepare__` are not defined.

**Comment by erictraut** [comment]
This behavior is too dynamic for a static type checker to understand. You will need to use other techniques if you want to use static type checking.
